### ISSUE TITLE
Linting via ESLint API fails with 'Cannot read property of undefined' but only on subsequent invocations

### ISSUE THREAD

**Issue by undeadcat** [typescript-eslint/typescript-eslint#498]
Possibly duplicate of #497, feel free to merge them.
I'm maintaining the ESLint integration on the side of Intellij IDEA-based IDEs and we've had several reports of intermittent 'Cannot read property of undefined' errors occurring, all when using 'typescript-eslint' (here's the latest one: https://youtrack.jetbrains.com/issue/WEB-38664). Unfortunately, nobody has shared enough of their project for us to reproduce the issue yet, so this is my attempt to construct a reproduction.
Apologies if this is a synthetic / unrealistic example.

* Checkout https://github.com/undeadcat/typescript-eslint-bug/tree/with-imports, do `npm install`.
(same code sample as #497, but `sourceType:module` is added to the .eslintrc and import declarations added to the .ts source files (this seems invalid, TS reports this as an error, but suppose we are in the middle of a refactoring, or *temporarily* misconfigured the project)).

* Linting from the console completes successfully:
```
node_modules/.bin/eslint ts.ts 

/Users/jetbrains/sources/testProjects/typescript-eslint-bug/ts.ts
   3:7   error    'SHOULD_BE_UNUSED' is assigned a value but never used  @typescript-eslint/no-unused-vars
  17:10  warning  Missing semicolon                                      semi

✖ 2 problems (1 error, 1 warning)
  0 errors and 1 warning potentially fixable with the `--fix` option.

```
* Linting via the ESLint API completes *only once but fails on subsequent invocations*

This case is important for us, because this is how an editor integration works. It creates a long-running process and then invokes ESLint via the API multiple times.

To reproduce, invoke `node ./lintFromApi.js` in the repository linked above.
For the first attempt, errors are reported correctly, subsequent attempts fail with
```stacktrace
TypeError: Cannot read property 'name' of undefined
    at Converter.convertNode (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:1402:28)
    at Converter.converter (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:66:27)
    at Converter.convertChild (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:138:21)
    at Converter.convertNode (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:1174:34)
    at Converter.converter (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:66:27)
    at Converter.convertChild (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:138:21)
    at nodes.map.statement (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:194:32)
    at Array.map (<anonymous>)
    at Converter.convertBodyExpressions (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:193:14)
    at Converter.convertNode (/Users/jetbrains/sources/testProjects/typescript-eslint-bug/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:427:32)
```

**Comment by bradzacher** [comment]
I checked out your repo, and it works fine for me:
```bash
bradzacher-mbp:typescript-eslint-bug-with-imports bradzacher$ node ./lintFromApi.js
attempt 0
errors and warnings: 
Missing semicolon.
'SHOULD_BE_UNUSED' is assigned a value but never used.
attempt 1
fatal error: {"ruleId":null,"fatal":true,"severity":2,"message":"Parsing error: Cannot read property 'name' of undefined"}
bradzacher-mbp:typescript-eslint-bug-with-imports bradzacher$ 
```

There might be something wrong with your packages somewhere? I'm not sure.

**Comment by bradzacher** [comment]
having a look here, the bug is caused by typescript not populating the `parent`.
We never handle that case in the code, because according to the types that shouldn't ever happen.

@uniqueiniquity - do you know why this is happening?

**Comment by bradzacher** [comment]
potentially related: https://github.com/microsoft/TypeScript/issues/24706

**Comment by uniqueiniquity** [comment]
@bradzacher good find, that is related.
Parent references _should_ always be set through the external API but aren't always set internally, so we must be working around that somehow. I'll take a look today and see what we might be missing.
